### PR TITLE
[MODEL] Add SINE model

### PR DESCRIPTION
**Pull request by Tokkiu**
Add sequential model SINE from Alibaba
The paper is https://arxiv.org/abs/2102.09267

### CODE REVIEW

**Comment by 2017pxy** [comment]
@Tokkiu Thx, we will review it as soon as possible.

**Review by ChangxinTian — CHANGES_REQUESTED** [review]
Thx.

File: recbole/model/sequential_recommender/sine.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            logits = torch.matmul(seq_output, test_item_emb.transpose(0, 1))
+            logits = F.log_softmax(logits, dim=1)
+            loss = self.loss_fct(logits, pos_items)
+            return loss + self.calculate_reg_loss() / 2
+
+    def calculate_reg_loss(self):
+        C_mean = torch.mean(self.C.weight, dim=1, keepdim=True)
+        C_reg = (self.C.weight - C_mean)
+        C_reg = C_reg.matmul(C_reg.T) / self.D
+        return (torch.norm(C_reg) ** 2 - torch.norm(torch.diag(C_reg)) ** 2) / 2
```
[the remark this answers is not in the document]
Comment (Tokkiu) [reply recbole/model/sequential_recommender/sine.py:111]: ok, I will fix them

**Review by ChangxinTian — CHANGES_REQUESTED** [review]
Thanks for your reformation.

File: recbole/model/sequential_recommender/sine.py [hunk 2]
```diff
@@ -153,19 +165,20 @@ def forward(self, item_seq, item_seq_len):
...
         x_u_re = x_u.unsqueeze(1).repeat(1, self.k, 1, 1)
-        mul_p_re = mul_p.unsqueeze(3).repeat(1, 1, 1, self.embedding_size)
+        mul_p_re = mul_p.unsqueeze(3)
         delta_k = x_u_re.mul(mul_p_re).sum(2)
         delta_k = F.normalize(delta_k, p=2, dim=2)
 
         # prototype sequence
         x_u_bar = P_k_t_b.matmul(C_u)
-        C_apt = F.softmax(torch.tanh(x_u_bar.matmul(self.w4)).matmul(self.w5), dim=1).reshape(-1, 1, self.max_seq_length).matmul(x_u_bar)
+        C_apt = F.softmax(torch.tanh(x_u_bar.matmul(self.w4)).matmul(self.w5), dim=1)
```
[the remark this answers is not in the document]
Comment (Tokkiu) [reply recbole/model/sequential_recommender/sine.py]: Thanks for your kind notice. I have changed them as same w3 as you can see in the latest commit.

**Review by ChangxinTian — APPROVED** [review]
Many thanks.

**Comment by 2017pxy** [comment]
@Tokkiu hi，firstly thanks for your contribution!

In order to test your model, we removed your code into v1.0.x branch and add some docs for it (in #1142). After the test, we will merge them into master branch from v1.0.x branch and this model will be released in next version (maybe v1.0.1).  Thus, this PR will be closed and you are welcome to review the new PR（#1142）if you have time. Many thanks!

**Comment by Tokkiu** [comment]
Dear Xingyu Pan,

Thank you for your kind notice. I think the code looks ok.

By the way, is there any thing or paper I can help to do further
implementation?

Best regards
Jingqi

Xingyu Pan ***@***.***> 于2022年2月20日周日 13:38写道：

**Comment by 2017pxy** [comment]
@Tokkiu Hi, we have a discussion called new model to be added (#611 ), you can find one you are interested in and implement it.

**Comment by 2017pxy** [comment]
This PR has been removed in #1142

**Pull request closed without merging**